text/0000-promise-based-initializers.md

```diff
@@ -0,0 +1,86 @@
+- Start Date: 2020-01-09
+- Relevant Team(s): Ember.js
+- RFC PR: https://github.com/emberjs/rfcs/pull/572
+- Tracking: (leave this empty)
+
+# Promise-based Initializers
+
+## Summary
+
+This RFC proposes the ability to return a promise from an initializer or instance-initializer
+and pause application boot until that promise resolves.
+
+## Motivation
+
+Currently, the earliest place that developers can write asynchronous code is in the ApplicationRoute.
+For some use cases, this is not early enough, and it is useful to be able to run asynchronous code
+before the application begins routing.
+
+For example, when booting an Ember application in an embedded environment, it may be necessary to
+listen for events from the environment to determine the initial state to determine the initial URL.
+
+Another example may be to interact with 3rd party libraries that need to be initialized early but
+have an asynchronous API.
+
+## Detailed Design
+
+The `Application` and `ApplicationInstance` (extending from `Engine` and `EngineInstance`
+classes respectively) are responsible for running initializers and instance initializers. Each
+implements a method that loops through the loaded initializers and executes the exported function.
+
+Initializer and Instance Initializer files each export an object with an `initialize` property
+that references a function. For backwards compatibility, an `async` property should be added, and
+should default to `false` if omitted. If true (or truthy), the Application/ApplicationInstance would
+expect a promise to be returned from the function and wait for the promise to resolve
+before running the next initializer in the queue. The value resolved from the promise would be
+ignored. Promises that reject would throw an exception in development and be caught in production
```

Comment:
> and be caught in production

What does this mean? If initialization of an application fails, I cannot reason about how to make that app functional again...

Reply:
it's up to do the developer to determine that. For example, we make a network request at the beginning to set some cookies. Without those cookies the app is perfectly functional, but could have unintended consequences later. Catching errors in production would just be a failsafe for apps who opts into this feature, but doesn't think through that. I'd be fine with removing this part.